base/path.jl

```diff
@@ -293,8 +296,15 @@ end
 
 else
 
-function joinpath(path::AbstractString, paths::AbstractString...)::String
-    for p in paths
+function joinpath(paths::Union{Tuple, AbstractVector})::String
+    assertstring(x) = x isa AbstractString || throw(ArgumentError("path component is not a string: $(repr(x))"))
+
+    isempty(paths) && return path_separator
```

Comment:
What is the rationale for returning `path_separator` here? Noting that `joinpath()` throws, and `joinpath("a") == "a"`, doesn't contain a path separator.

Reply:
When extending the interface to collections I had to pick something for `[]` and `()`.
@StefanKarpinski stated clearly in #20912 that `joinpath` should not throw, I treat these errors as typechecking. As to why separator:
* Separator on both Windows and Unix systems is a correct path to the root of the current drive. 
* Empty string is not a valid path in Unix.
* Root of the current drive (`/` on Unix and `C:\` on Windows) require looking at filesystem on Windows, then `joinpath` would behave differently depending on current working directory and would no longer be just a data manipulation function (https://github.com/JuliaLang/julia/issues/20912#issuecomment-284441651).

I don't have any better idea.

Reply:
Ok. I interpret (I can be wrong!) "joinpath should not throw" as it shouldn't depend on the content of the string arguments. But `joinpath()` already throws (a `MethodError`), and I wouldn't be too worried that `joinpath(())` also throws, and we keep the equivalence `joinpath(a::Vector{String}) == joinpath(a...)`. 
And you could define only one method taking string arguments as
```
joinpath(paths::AbstractString...)::String = joinpath(paths)
```
I'm not sure if I would prefer `joinpath()` or `joinpath([])` to throw or to return something which is "neutral" for path joining, which would rather be the empty string than `/` (returning the empty string is not the same as throwing an error, even if the empty string is not a valid path).

Reply:
After some consideration and @rfourquet's comments I now support throwing here:
* this was the original author's intent, which can be seen in missing `joinpath()`,
* there doesn't seem to be any clear best candidate for return value,
* symmetry with `joinpath(AbstractString...)`, which is now implemented as
```julia
joinpath(paths::AbstractString...)::String = joinpath(paths),
``` 
* `joinpath` doesn't seem to be used in many dynamic situations where `[]` might come up, it's usually applied to a pair of some directory variable and a filename string.

I modified the code accordingly.

Reply:
I see that you updated the code according to my feedback, but again, my opinion is just my opinion :) Maybe @StefanKarpinski will have another take, let's wait for more feedback.